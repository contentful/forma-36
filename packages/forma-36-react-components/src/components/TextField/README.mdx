---
title: 'TextField'
type: 'component'
status: 'stable'
slug: /components/text-field/
github: 'https://github.com/contentful/forma-36/tree/master/packages/forma-36-react-components/src/components/TextField'
storybook: 'https://f36-storybook.contentful.com/?path=/story/components-textfield--default'
---

`TextField` is a composition of the `FormLabel`, `TextInput`, `HelpText` and `ValidationMessage` components.
It allows users to input text or numbers and should be used in a form context.

## Table of contents

- [How to use TextField](#how-to-use-TextField)
- [Code examples](#code-examples)
- [Code examples](#content-recommendations)
- [Accessibility](#accessibility)

## How to use TextField

The TextField component can be configured in a number of different ways. Use this guidance to determine the correct variation based on the context.

## Code examples

```jsx
<React.Fragment>
  <TextField
    labelText="Email address"
    helpText="Please enter your email address"
    required
  />
</React.Fragment>
```

## TextField variations

### Compositional properties

Since the TextField is a composition of multiple components, it is possible to pass child component specific properties.
For this purpose, the TextField exposes the following properties:

```jsx
<React.Fragment>
  <TextField
    labelText="E-mail address"
    helpText="Please enter your email address"
    textInputProps={{ withCopyButton: true }}
    textLinkProps={{ text: 'privacy' }}
    formLabelProps={{ requiredText: 'required' }}
    required
  />
</React.Fragment>
```

### Label and text properties (`labelText, helpText, validationMessage`)

```jsx
<React.Fragment>
  <TextField
    labelText="E-mail address"
    helpText="Please enter your email address"
    validationMessage="This address is not valid"
  />
</React.Fragment>
```

### Required field (`required`)

```jsx
<React.Fragment>
  <TextField
    labelText="E-mail address"
    helpText="Please enter your email address"
    required
  />
</React.Fragment>
```

### Render textarea (`textfield`)

Instead of an input, the TextField component can render a text area for long text requirements.

```jsx
<React.Fragment>
  <TextField
    labelText="Description"
    helpText="Please enter a description"
    textarea
  />
</React.Fragment>
```

### Additional TextLink (`textLinkProps`)

The TextField can be configured to display an additional TextLink in the label.

```jsx
<React.Fragment>
  <TextField
    labelText="User Name"
    textLinkProps={{
      icon: 'Lock',
      text: 'Unlock to edit',
      onClick: (e) => {
        console.log('textLinkClick', e);
      },
    }}
    value="Lorem Ipsum"
    textInputProps={{
      disabled: true,
    }}
  />
</React.Fragment>
```

### Character count (`countCharacters`)

The TextField can be configured to display a character count. This is useful when an input field has a character limit.<br/>
**Note**: The `maxLength` property must be provided in the TextInput component using `textInputProps`.

```jsx
<React.Fragment>
  <TextField
    labelText="User Name"
    countCharacters
    value="Lorem Ipsum"
    textInputProps={{ maxLength: 20 }}
  />
</React.Fragment>
```

## Content recommendations

- Use direct, succinct copy that helps the user to successfully complete the form
- Do not preface the instructions with introductory text, such as "please"
- To add additional context, link out to documentation
- In cases where additional context is unnecessary, help text can be substituted for text in the label
- Do not use punctuation for labels, help text and links
- Use sentence style caps (in which only the first word of a sentence or term is capitalized)

## Accessibility

- To ensure `TextField` meets accessibility standards, provide `labelText`, `name` and `id`
