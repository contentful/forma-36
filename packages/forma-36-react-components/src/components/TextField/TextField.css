@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';
@import 'settings/dimensions';

.TextField {
  display: block;
}

.TextField--small {
  width: var(--input-width-small);
}

.TextField--medium {
  width: var(--input-width-medium);
}

.TextField--large {
  width: var(--input-width-large);
}

.TextField--full {
  width: 100%;
}

.TextField__label-wrapper {
  display: flex;
  font-size: var(--font-size-m);
  align-items: center;
}

.TextField__label-link {
  margin-left: auto;
  margin-bottom: calc(1rem * (8 / var(--font-base-default)));
}

.TextField__count {
  flex-grow: 1;
  text-align: right;
}
