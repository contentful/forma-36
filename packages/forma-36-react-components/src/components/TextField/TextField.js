import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import ValidationMessage from '../ValidationMessage';
import FormLabel from '../FormLabel';
import HelpText from '../HelpText';
import TextInput from '../TextInput';
import TextLink from '../TextLink';
import Textarea from '../Textarea';
import styles from './TextField.css';

class TextField extends React.Component {
  static propTypes = {
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    validationMessage: PropTypes.string,
    testId: PropTypes.string,
    extraClassNames: PropTypes.string,
    formLabelProps: PropTypes.object,
    textLinkProps: PropTypes.object,
    textInputProps: PropTypes.object,
    name: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    labelText: PropTypes.string,
    helpText: PropTypes.string,
    required: PropTypes.bool,
    textarea: PropTypes.bool,
    countCharacters: PropTypes.bool,
    onChange: PropTypes.func,
    onBlur: PropTypes.func,
  };

  static defaultProps = {
    validationMessage: undefined,
    extraClassNames: undefined,
    testId: 'cf-ui-text-field',
    formLabelProps: undefined,
    textLinkProps: undefined,
    textInputProps: undefined,
    labelText: undefined,
    onChange: undefined,
    helpText: undefined,
    onBlur: undefined,
    value: undefined,
    textarea: false,
    required: false,
    countCharacters: false,
  };

  state = { value: this.props.value };

  // Store a copy of the value in state.
  // This is used by this component when the `countCharacters`
  // option is on
  handleOnChange = evt => {
    this.setState({ value: evt.target.value });
    if (this.props.onChange) this.props.onChange(evt);
  };

  render() {
    const {
      validationMessage,
      extraClassNames,
      textInputProps,
      testId,
      formLabelProps,
      textLinkProps,
      labelText,
      helpText,
      textarea,
      countCharacters,
      required,
      onChange,
      onBlur,
      value,
      name,
      id,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.TextField, extraClassNames);

    const Element = textarea ? Textarea : TextInput;
    return (
      <div className={classNames} {...otherProps} data-test-id={testId}>
        <div className={styles['TextField__label-wrapper']}>
          <FormLabel {...{ ...formLabelProps, htmlFor: id, required }}>
            {labelText}
          </FormLabel>
          {textLinkProps && (
            <TextLink
              {...{
                ...textLinkProps,
                extraClassNames: styles['TextField__label-link'],
              }}
            >
              {textLinkProps.text}
            </TextLink>
          )}
        </div>
        <Element
          {...{
            error: !!validationMessage,
            name,
            id,
            onBlur,
            onChange: this.handleOnChange,
            value,
            required,
            ...textInputProps,
          }}
        />
        {validationMessage && (
          <ValidationMessage
            extraClassNames={styles['TextField__validation-message']}
          >
            {validationMessage}
          </ValidationMessage>
        )}
        {(helpText || countCharacters) && (
          <div className={styles.TextField__hints}>
            {helpText && (
              <HelpText extraClassNames={styles['TextField__help-text']}>
                {helpText}
              </HelpText>
            )}
            {countCharacters && textInputProps && textInputProps.maxLength && (
              <HelpText
                extraClassNames={cn(
                  styles['TextField__help-text'],
                  styles.TextField__count,
                )}
              >
                {this.state.value ? this.state.value.length : 0}/
                {textInputProps.maxLength}
              </HelpText>
            )}
          </div>
        )}
      </div>
    );
  }
}

export default TextField;
