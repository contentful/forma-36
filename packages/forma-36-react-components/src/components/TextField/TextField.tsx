import React, {
  useState,
  ChangeEvent,
  ChangeEventHandler,
  FocusEventHandler,
  ElementType,
} from 'react';
import cn from 'classnames';
import ValidationMessage from '../ValidationMessage';
import FormLabel, { FormLabelProps } from '../FormLabel';
import HelpText from '../HelpText';
import TextInput, { TextInputProps } from '../TextInput';
import TextLink, { TextLinkProps } from '../TextLink';
import Textarea, { TextareaProps } from '../Textarea';
import styles from './TextField.css';

export interface TextFieldProps {
  name: string;
  id: string;
  width?: string;
  labelText: string;
  value?: string;
  validationMessage?: string;
  testId?: string;
  className?: string;
  formLabelProps?: Partial<FormLabelProps>;
  textLinkProps?: Partial<TextLinkProps>;
  textInputProps?: Partial<TextInputProps> | Partial<TextareaProps>;
  helpText?: string;
  required?: boolean;
  textarea?: boolean;
  countCharacters?: boolean;
  onChange?: ChangeEventHandler;
  onBlur?: FocusEventHandler;
  onFocus?: FocusEventHandler;
}

export interface TextFieldState {
  value?: string;
  initialValue?: string;
}

const defaultProps: Partial<TextFieldProps> = {
  testId: 'cf-ui-text-field',
  textarea: false,
  required: false,
  countCharacters: false,
  width: 'full',
};

export const TextField = (props: TextFieldProps) => {
  const {
    validationMessage,
    className,
    textInputProps,
    testId,
    width,
    formLabelProps,
    textLinkProps,
    labelText,
    helpText,
    textarea,
    countCharacters,
    required,
    onChange,
    onBlur,
    onFocus,
    value,
    name,
    id,
    ...otherProps
  } = props;
  const [valueState, setValueState] = useState<string | undefined>(value);

  // Store a copy of the value in state.
  // This is used by this component when the `countCharacters`
  // option is on
  const handleOnChange = (
    e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>,
  ) => {
    setValueState(e.target.value);
    if (onChange) onChange(e);
  };

  const widthClass = `TextField--${width}`;
  const classNames = cn(styles['TextField'], styles[widthClass], className);

  const Element: ElementType = textarea ? Textarea : TextInput;

  return (
    <div className={classNames} {...otherProps} data-test-id={testId}>
      <div className={styles['TextField__label-wrapper']}>
        <FormLabel {...formLabelProps} htmlFor={id} required={required}>
          {labelText}
        </FormLabel>
        {textLinkProps && (
          <TextLink
            {...{
              ...textLinkProps,
              className: styles['TextField__label-link'],
            }}
          >
            {textLinkProps.text}
          </TextLink>
        )}
      </div>
      <Element
        {...{
          error: !!validationMessage,
          name,
          id,
          onBlur,
          onFocus,
          onChange: handleOnChange,
          value,
          required,
          ...textInputProps,
          width: 'full',
        }}
      />
      {validationMessage && (
        <ValidationMessage className={styles['TextField__validation-message']}>
          {validationMessage}
        </ValidationMessage>
      )}
      {(helpText || countCharacters) && (
        <div className={styles['TextField__hints']}>
          {helpText && (
            <HelpText className={styles['TextField__help-text']}>
              {helpText}
            </HelpText>
          )}
          {countCharacters && textInputProps && textInputProps.maxLength && (
            <HelpText
              className={cn(
                styles['TextField__help-text'],
                styles['TextField__count'],
              )}
            >
              {valueState ? valueState.length : 0}/{textInputProps.maxLength}
            </HelpText>
          )}
        </div>
      )}
    </div>
  );
};

TextField.defaultProps = defaultProps;

export default TextField;
