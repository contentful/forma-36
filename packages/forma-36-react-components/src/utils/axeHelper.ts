import { configureAxe } from 'jest-axe';

const axe = configureAxe({
  rules: {
    region: { enabled: false },
  },
});

export default axe;
