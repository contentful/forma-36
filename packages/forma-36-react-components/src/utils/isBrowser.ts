const isBrowser = () => typeof window !== 'undefined';

export default isBrowser;
