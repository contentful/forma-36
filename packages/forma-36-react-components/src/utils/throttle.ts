const throttle = (delay = 200, fn: Function) => {
  let lastCall = 0;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const throttleExec = (...args: any[]) => {
    const now = new Date().getTime();
    if (now - lastCall < delay) {
      return;
    }
    lastCall = now;
    return fn(...args); // eslint-disable-line
  };
  return throttleExec;
};

export default throttle;
