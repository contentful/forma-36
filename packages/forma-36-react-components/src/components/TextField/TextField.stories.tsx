import React from 'react';

import TextField, { TextFieldProps } from './TextField';
import SectionHeading from '../Typography/SectionHeading';
import Flex from '../Flex/Flex';
import Grid from '../Grid/Grid';

export default {
  title: 'Form Elements/TextField',
  component: TextField,
  parameters: {
    propTypes: [TextField['__docgenInfo']],
  },
  argTypes: {
    className: { control: { disable: true } },
    testId: { control: { disable: true } },
    isReadOnly: { control: { disable: true } },
    onChange: { control: { disable: true } },
    onBlur: { control: { disable: true } },
    onCopy: { control: { disable: true } },
    onFocus: { control: { disable: true } },
    onKeyPress: { control: { disable: true } },
    textLinkProps: { control: { disable: true } },
    textInputProps: { control: { disable: true } },
    formLabelProps: { control: { disable: true } },
  },
};

interface Args extends TextFieldProps {
  placeholder?: string;
  maxLength?: number;
  disabled?: boolean;
}

export const Basic = ({ placeholder, maxLength, disabled, ...args }: Args) => (
  <TextField {...args} textInputProps={{ placeholder, maxLength, disabled }} />
);

Basic.args = {
  id: 'basic-input',
  name: 'basic-input',
  labelText: 'Label text',
  placeholder: 'Placeholder text',
  maxLength: 100,
  disabled: false,
};

export const WithTextarea = ({
  placeholder,
  maxLength,
  disabled,
  ...args
}: Args) => (
  <TextField {...args} textInputProps={{ placeholder, maxLength }} textarea />
);

WithTextarea.args = {
  ...Basic.args,
};

export const WithTextLink = ({
  placeholder,
  maxLength,
  disabled,
  textLinkProps,
  ...args
}: Args) => (
  <TextField
    {...args}
    textInputProps={{ placeholder, maxLength }}
    textLinkProps={textLinkProps}
    textarea
  />
);

WithTextLink.args = {
  textLinkProps: {
    icon: 'Lock',
    text: 'Unlock to edit',
  },
  ...Basic.args,
};

export const Overview = (args: TextFieldProps) => (
  <Flex style={{ minWidth: '800px' }} flexDirection="column">
    <Flex flexDirection="column" marginBottom="spacingXl">
      <Flex marginBottom="spacingS">
        <SectionHeading element="h3">TextField default</SectionHeading>
      </Flex>
      <Grid columns="1fr 1fr" columnGap="spacingXl">
        <TextField {...args} />
        <TextField {...args} textarea />
      </Grid>
    </Flex>

    <Flex flexDirection="column" marginBottom="spacingXl">
      <Flex marginBottom="spacingS">
        <SectionHeading element="h3">
          TextField with countCharacters
        </SectionHeading>
      </Flex>
      <Grid columns="1fr 1fr" columnGap="spacingXl">
        <TextField
          {...args}
          countCharacters
          textInputProps={{ maxLength: 50 }}
        />
        <TextField
          {...args}
          countCharacters
          textInputProps={{ maxLength: 50 }}
          textarea
        />
      </Grid>
    </Flex>

    <Flex flexDirection="column" marginBottom="spacingXl">
      <Flex marginBottom="spacingS">
        <SectionHeading element="h3">
          TextField with default value
        </SectionHeading>
      </Flex>
      <Grid columns="1fr 1fr" columnGap="spacingXl">
        <TextField value="Example value" {...args} />
        <TextField value="Example value" {...args} textarea />
      </Grid>
    </Flex>

    <Flex flexDirection="column" marginBottom="spacingXl">
      <Flex marginBottom="spacingS">
        <SectionHeading element="h3">TextField with help text</SectionHeading>
      </Flex>
      <Grid columns="1fr 1fr" columnGap="spacingXl">
        <TextField {...args} helpText="help text" />
        <TextField {...args} helpText="help text" textarea />
      </Grid>
    </Flex>

    <Flex flexDirection="column" marginBottom="spacingXl">
      <Flex marginBottom="spacingS">
        <SectionHeading element="h3">
          TextField with validation message
        </SectionHeading>
      </Flex>
      <Grid columns="1fr 1fr" columnGap="spacingXl">
        <TextField {...args} validationMessage="Validation message" />
        <TextField {...args} validationMessage="Validation message" textarea />
      </Grid>
    </Flex>

    <Flex flexDirection="column">
      <Flex marginBottom="spacingS">
        <SectionHeading element="h3">TextField disabled</SectionHeading>
      </Flex>
      <Grid columns="1fr 1fr" columnGap="spacingXl">
        <TextField
          {...args}
          textInputProps={{
            disabled: true,
          }}
        />
        <TextField
          {...args}
          textInputProps={{
            disabled: true,
          }}
          textarea
        />
      </Grid>
    </Flex>
  </Flex>
);

Overview.args = {
  ...Basic.args,
};
