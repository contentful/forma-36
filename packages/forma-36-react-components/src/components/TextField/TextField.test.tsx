import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Lock } from '@contentful/f36-icons';

import { axe } from '../../utils/axeHelper';
import { TextField } from './TextField';

it('renders the component', () => {
  const { container } = render(
    <TextField labelText="test" name="someComponent" id="someComponent" />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      className="my-extra-class"
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with validation message', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      validationMessage="Field is required"
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with required text', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      formLabelProps={{
        requiredText: 'required',
      }}
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with help text', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      helpText="You need to fill out this field"
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with a textlink', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      textLinkProps={{
        icon: Lock,
        text: 'Unlock to edit',
        onClick: () => {},
      }}
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with a value', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      value="1234"
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('updates the value', () => {
  const { container } = render(
    <TextField
      id="someComponent"
      name="someComponent"
      labelText="test"
      data-test-id="input"
      value=""
    />,
  );

  const input = container.querySelector(
    '[data-test-id="cf-ui-text-field"] input',
  );

  userEvent.type(input, 'new value');

  expect(screen.getByDisplayValue('new value')).toBeInTheDocument();
  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with as textarea', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      textarea
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with a copybutton', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      textInputProps={{
        withCopyButton: true,
      }}
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with characters count', () => {
  const { container } = render(
    <TextField
      countCharacters
      labelText="test"
      name="someComponent"
      id="someComponent"
      textInputProps={{
        maxLength: 20,
      }}
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('updates the characters counter', () => {
  render(
    <TextField
      countCharacters
      value="foo"
      labelText="test"
      name="someComponent"
      id="someComponent"
      textInputProps={{
        maxLength: 20,
      }}
    />,
  );

  const counter = screen.getByText('3/20');

  userEvent.type(screen.getByLabelText('test'), 'bar');
  expect(counter.textContent).toBe('6/20');
});

it('renders the component with small width', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      width="full"
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with max length of characters', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      textInputProps={{
        maxLength: 10,
      }}
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with a placeholder text', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      textInputProps={{
        placeholder: 'placeholder',
      }}
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with a rows defined', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      textInputProps={{
        rows: 2,
      }}
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with additional prop', () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      textInputProps={{
        testId: 'test',
      }}
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const { container } = render(
    <TextField
      labelText="test"
      name="someComponent"
      id="someComponent"
      className="my-extra-class"
    />,
  );
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});
