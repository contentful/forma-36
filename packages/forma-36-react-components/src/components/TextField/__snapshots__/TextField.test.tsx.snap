// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      name="someComponent"
      value=""
    />
  </div>
</div>
`;

exports[`renders the component with a copybutton 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      name="someComponent"
      value=""
    />
    <div
      class="CopyButton TextInput__copy-button"
      data-test-id="cf-ui-copy-button"
      id="copyButton"
    >
      <span>
        <button
          class="CopyButton__button"
          type="button"
        >
          <span
            class="TabFocusTrap CopyButton__TabFocusTrap"
            tabindex="-1"
          >
            <span
              class="CopyButton__text"
            >
              Copy 
               to clipboard
            </span>
            <svg
              class="Icon Icon--small Icon--muted"
              data-test-id="cf-ui-icon"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
              />
            </svg>
          </span>
        </button>
      </span>
    </div>
  </div>
</div>
`;

exports[`renders the component with a placeholder text 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      name="someComponent"
      placeholder="placeholder"
      value=""
    />
  </div>
</div>
`;

exports[`renders the component with a rows defined 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      name="someComponent"
      rows="2"
      value=""
    />
  </div>
</div>
`;

exports[`renders the component with a textlink 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
    <button
      class="TextLink TextField__label-link TextLink--primary"
      data-test-id="cf-ui-text-link"
      type="button"
    >
      <span
        class="TabFocusTrap"
        tabindex="-1"
      >
        <div
          class="TextLink__icon-wrapper"
        >
          <svg
            class="Icon Icon--small Icon--primary TextLink__icon"
            data-test-id="cf-ui-icon"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
            />
          </svg>
        </div>
        Unlock to edit
      </span>
    </button>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      name="someComponent"
      value=""
    />
  </div>
</div>
`;

exports[`renders the component with a value 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      name="someComponent"
      value="1234"
    />
  </div>
</div>
`;

exports[`renders the component with additional prop 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="test"
      id="someComponent"
      name="someComponent"
      value=""
    />
  </div>
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<div
  class="TextField TextField--full my-extra-class"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      name="someComponent"
      value=""
    />
  </div>
</div>
`;

exports[`renders the component with as textarea 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="Textarea Textarea--full"
  >
    <textarea
      aria-label="someComponent"
      class="Textarea__textarea"
      data-test-id="cf-ui-textarea"
      id="someComponent"
      name="someComponent"
    />
  </div>
</div>
`;

exports[`renders the component with characters count 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      maxlength="20"
      name="someComponent"
      value=""
    />
  </div>
  <div
    class="TextField__hints"
  >
    <p
      class="HelpText TextField__help-text TextField__count"
      data-test-id="cf-ui-help-text"
    >
      0
      /
      20
    </p>
  </div>
</div>
`;

exports[`renders the component with help text 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      name="someComponent"
      value=""
    />
  </div>
  <div
    class="TextField__hints"
  >
    <p
      class="HelpText TextField__help-text"
      data-test-id="cf-ui-help-text"
    >
      You need to fill out this field
    </p>
  </div>
</div>
`;

exports[`renders the component with max length of characters 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      maxlength="10"
      name="someComponent"
      value=""
    />
  </div>
</div>
`;

exports[`renders the component with required text 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      name="someComponent"
      value=""
    />
  </div>
</div>
`;

exports[`renders the component with small width 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      name="someComponent"
      value=""
    />
  </div>
</div>
`;

exports[`renders the component with validation message 1`] = `
<div
  class="TextField TextField--full"
  data-test-id="cf-ui-text-field"
>
  <div
    class="TextField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="someComponent"
    >
      test
    </label>
  </div>
  <div
    class="TextInput TextInput--full TextInput--negative"
  >
    <input
      aria-label="someComponent"
      class="TextInput__input"
      data-test-id="cf-ui-text-input"
      id="someComponent"
      name="someComponent"
      value=""
    />
  </div>
  <div
    class="ValidationMessage TextField__validation-message"
    data-test-id="cf-ui-validation-message"
  >
    <svg
      class="Icon Icon--small Icon--negative ValidationMessage__icon"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
      />
    </svg>
    <p
      class="ValidationMessage__text"
    >
      Field is required
    </p>
  </div>
</div>
`;
